### PR TITLE
feat: add retention policy to  gcs module [DA-3815] 

### PR DESCRIPTION
**Pull request by thadchais**
### Pull Request Submission Checklist

Please confirm that you have done the following before requesting reviews:

- [x] I have confirmed that the PR type is appropriate for the change I am making according to the [Honest Pull Request and Commit Message Naming Conventions](https://www.notion.so/honestbank/Pull-Request-and-Commit-Message-Naming-Conventions-bd97f2cbb34c4c73b1ff3a3e384b850c).
- [x] I have typed an adequate description that explains **why** I am making this change.
- [x] I have installed and run standard pre-commit hooks that lints and validates my code.
- [ ] All entities that I am working with are up-to-date in Backstage; if updates are needed, I have linked the relevant PRs. [Backstage guide](https://www.notion.so/honestbank/How-to-Write-a-Backstage-Service-Catalog-Entry-a-catalog-info-yaml-file-21845ff72e404b14aed2ac989fb202cf?pvs=4)

## Description

This PR introduces the following changes to the gcp_gcs_bucket module:  
* Retention Lock Configuration:  
Added a retention_policy block to the google_storage_bucket resource to enable retention lock on the bucket.
Introduced two new variables: retention_lock_enabled and retention_lock_duration_seconds to control the retention lock settings.

* Object Versioning:

Added a new variable: object_versioning_enabled to control the versioning of objects in the bucket.

## Why

Retention Lock: Ensures that objects in the bucket are protected from deletion for a specified period, enhancing data security and compliance with data retention policies.

Data Engineers use this feature to backup data platform and secure the archived data. Here is the [design document](https://www.notion.so/honestbank/Data-Lake-Backup-and-Restoration-GCS-buckets-820f9a6d7118459f9ad0d3ac7dbb8abd?pvs=4)

Object Versioning: Allows multiple versions of an object to be stored, enabling retrieval, restoration, or permanent deletion of specific versions of objects.

## APPENDIX
The variable `var.object_versioning_enabled` is a boolean variable that, when set to `true`, enables versioning for the Google Cloud Storage bucket. This means that multiple versions of an object can be stored in the bucket, allowing you to retrieve, restore, or permanently delete specific versions of objects.

### Mutually Exclusive with Retention Lock

The `retention_policy` block, which includes the `retention_lock_enabled` and `retention_lock_duration_seconds` variables, is used to enforce a retention lock on the bucket. This lock ensures that objects cannot be deleted or replaced for a specified period.

However, the `retention_policy` and object versioning are mutually exclusive. This means you cannot enable both at the same time for a single bucket. If `retention_lock_enabled` is set to `true`, you cannot enable `object_versioning_enabled`, and vice versa. This restriction is in place because the retention policy enforces strict deletion rules that conflict with the flexibility provided by object versioning.

More detail: [Considerations with other features](https://cloud.google.com/storage/docs/bucket-lock#retention-interactions)

Completed DA-3820 fixed Terratest workflow

- - -
This change is [Reviewable](https://reviewable.io/reviews/honestbank/terraform-gcp-gcs/29)

### CODE REVIEW

File: modules/gcp_gcs_bucket/inputs.tf [hunk 1]
```diff
@@ -56,6 +56,24 @@ variable "lifecycle_rules" {
...
+variable "retention_lock_enabled" {
+  type        = bool
+  description = "If set to true, the bucket will be locked and objects in the bucket will be protected from deletion. Note that retention_policy cannot be used with object versioning. They are mutually exclusive."
+  default     = false
+}
+
+variable "retention_lock_duration_seconds" {
+  type        = number
+  description = "The duration in seconds that objects in the bucket must be retained and cannot be deleted or replaced. The value must be in between 604800(7 days) and 31536000(365 days)."
+  default     = 86400 # 1 day
```
Comment (olli-honest) [inline modules/gcp_gcs_bucket/inputs.tf:74]: The line above says:

> The value must be in between 604800(7 days) and 31536000(365 days).

**Review by olli-honest — COMMENTED** [review]
Should this be covered by the test case as well [here](https://github.com/honestbank/terraform-gcp-gcs/blob/main/examples/bucket/main.tf)?

File: modules/gcp_gcs_bucket/inputs.tf (hunk 1 above)
Comment (olli-honest) [reply modules/gcp_gcs_bucket/inputs.tf:74]: https://developer.hashicorp.com/terraform/language/values/variables#custom-validation-rules

Comment (ChristianWitts) [reply modules/gcp_gcs_bucket/inputs.tf:74]: You can add a validation block so a plan will fail if you provide a value that is out-of-bounds, for example:

```hcl
variable "retention_lock_duration_seconds" {
  ...

  validation {
    condition     = var.retention_lock_enabled == true && (var. retention_lock_duration_seconds >= 604800 && var. retention_lock_duration_seconds <= 31536000)
    error_message = "The retention_lock_duration_seconds must be between 604800(7 days) and 31536000(365 days)."
  }
}
```

**Review by MXfive — DISMISSED** [review]

**Review by ChristianWitts — DISMISSED** [review]

**Review by olli-honest — APPROVED** [review]

**Review by MXfive — APPROVED** [review]

**Review by olli-honest — APPROVED** [review]

**Pull request merged**